xpti/include/xpti/xpti_trace_framework.hpp

```diff
@@ -479,84 +559,930 @@ inline std::string readMetadata(const metadata_t::value_type &MD) {
+                 uint16_t traceType = (uint16_t)
+                     xpti::trace_point_type_t::function_begin)
+      : MTraceEvent(traceEvent), MUserData(UserData), MStreamId(streamId),
+        MTraceType(traceType), MScopedNotify(false), MTraceEnabled(false) {
+    // Reduce calls to xptiCheckTraceENabled() by caching it
+    MTraceEnabled = xptiCheckTraceEnabled(MStreamId, MTraceType);
+    if (!MTraceEnabled)
+      return;
+
+    MCorrelationId = xptiGetUniqueId();
   }
-  ~stash_tuple() {
-    if (m_stashed) {
-      xptiUnstashTuple();
-    }
+
+  /// @brief Sends a notification to subscribers.
+  ///
+  /// This method checks if tracing is enabled for the stream and trace type.
+  /// If tracing is enabled, it sends a notification to subscribers. This call
+  /// is used for trace types that do not have a begin and end scopes.
+  ///
+  /// @return A reference to this object.
+  ///
+  notify_scope_t &notify() {
+    if (!MTraceEnabled)
+      return *this;
+
+    uint16_t TraceType = MTraceType & 0xfffe;
+    xptiNotifySubscribers(MStreamId, TraceType, nullptr, MTraceEvent,
+                          MCorrelationId, static_cast<const void *>(MUserData));
+    return *this;
+  }
+
+  /// @brief Sends a scoped notification to subscribers.
+  ///
+  /// This method checks if tracing is enabled for the stream and trace type.
+  /// If tracing is enabled, it sends a scoped notification to subscribers and
+  /// sets the scoped notify flag so the end of scope notification can be sent
+  /// out when the object is destroyed.
+  ///
+  /// @return A reference to this object.
+  ///
+  notify_scope_t &scopedNotify() {
+    if (!MTraceEnabled)
+      return *this;
+
+    uint16_t TraceType = MTraceType & 0xfffe;
+    MScopedNotify = true;
+    xptiNotifySubscribers(MStreamId, TraceType, nullptr, MTraceEvent,
+                          MCorrelationId, static_cast<const void *>(MUserData));
+    return *this;
+  }
+
+  /// @brief The destructor of the notify_scope_t class.
+  ///
+  /// The destructor checks if tracing is enabled for the stream and trace type.
+  /// If tracing is enabled, it sends an end-of-scope notification to
+  /// subscribers.
+  ///
+  ~notify_scope_t() {
+    if (!MScopedNotify || !MTraceEnabled)
+      return;
+
+    uint16_t TraceType = MTraceType | 1;
+    xptiNotifySubscribers(MStreamId, TraceType, nullptr, MTraceEvent,
+                          MCorrelationId, static_cast<const void *>(MUserData));
   }
 
 private:
-  bool m_stashed;
+  /// The trace event for which the notifications are being sent out
+  xpti::trace_event_data_t *MTraceEvent;
+  /// The user data to be sent out with the notification
+  const char *MUserData;
+  /// The stream with which the notification is associated
+  uint8_t MStreamId;
+  /// The type of the trace event: function_begin, signal, etc
+  uint16_t MTraceType;
+  /// The correlation ID for the notification begin/end scope calls
+  uint64_t MCorrelationId;
```

Comment:
not initialized in ctor

Reply:
MTraceType is, fixed MCorrelationId only to fix warnings as it is actually set in the constructor.